app/org/sagebionetworks/bridge/services/ParticipantService.java

```diff
@@ -280,40 +320,29 @@ public StudyParticipant getParticipant(Study study, Account account, boolean inc
-            List<Subpopulation> subpopulations = subpopService.getSubpopulations(study.getStudyIdentifier(), false);
-            for (Subpopulation subpop : subpopulations) {
-                // always returns a list, even if empty
-                List<UserConsentHistory> history = getUserConsentHistory(account, subpop.getGuid());
-                consentHistories.put(subpop.getGuidString(), history);
-            }
-            builder.withConsentHistories(consentHistories);
-            // To calculate consent status, we need construct a CriteriaContext from RequestInfo (only for the 
-            // information we don't store in account).
-            RequestInfo requestInfo = cacheProvider.getRequestInfo(account.getId());
-            if (requestInfo != null) {
-                CriteriaContext criteriaContext = new CriteriaContext.Builder()
-                        .withStudyIdentifier(study.getStudyIdentifier())
-                        .withUserId(account.getId())
-                        .withHealthCode(account.getHealthCode())
-                        .withClientInfo(requestInfo.getClientInfo())
-                        .withLanguages(requestInfo.getLanguages())
-                        .withUserDataGroups(account.getDataGroups())
-                        .withUserSubstudyIds(assoc.getSubstudyIdsVisibleToCaller())
-                        .build();
-                Map<SubpopulationGuid, ConsentStatus> consentStatusMap = consentService.getConsentStatuses(
-                        criteriaContext, account);
-                boolean isConsented = ConsentStatus.isUserConsented(consentStatusMap);
-                builder.withConsented(isConsented);
-            }
-        }
-        return builder.build();
+        return builder;
+    }
+    
+    private StudyParticipant.Builder addHistory(StudyParticipant.Builder builder, Account account, CriteriaContext context) {
+        Map<String,List<UserConsentHistory>> consentHistories = Maps.newHashMap();
+        // The history includes all subpopulations whether they match the user or not.
+        List<Subpopulation> subpopulations = subpopService.getSubpopulations(context.getStudyIdentifier(), false);
+        for (Subpopulation subpop : subpopulations) {
+            // always returns a list, even if empty
+            List<UserConsentHistory> history = getUserConsentHistory(account, subpop.getGuid());
+            consentHistories.put(subpop.getGuidString(), history);
+        }
+        builder.withConsentHistories(consentHistories);
+        return builder;
+    }
+    
+    private StudyParticipant.Builder addConsentStatus(StudyParticipant.Builder builder, Account account, CriteriaContext context) {
+        Map<SubpopulationGuid, ConsentStatus> consentStatuses = consentService.getConsentStatuses(context, account);
+        boolean isConsented = ConsentStatus.isUserConsented(consentStatuses);
+        builder.withConsented(isConsented);
+        return builder;
```

Comment:
This is called a second time when we create a session; it would be good to avoid that.

Reply:
I'm going to need to refactor this so consent statuses are set, not the flag, and then the flag can be calculated from within the session. As a consequent, this expensive calculation should not need to be redone in the session code. Filed BRIDGE-2419 on this.